### PR TITLE
CI: enhance with go check and escapes detect

### PR DESCRIPTION
**Pull request by SamYuan1990**
Signed-off-by: Sam Yuan <yy19902439@126.com>

### CODE REVIEW

**Comment by SamYuan1990** [comment]
```
# github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/registry
Error: pkg/adaptor/hypervisor/registry/register.go:9:9: undefined: newServer
make: *** [Makefile:59: vet] Error 2
```

I am confused a little bit for where is the newServer function?

**Comment by SamYuan1990** [comment]
memory escapes can be found at https://github.com/SamYuan1990/cloud-api-adaptor/actions/runs/3321361525/jobs/5488906203

**Comment by SamYuan1990** [comment]
> > ```
> > # github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/registry
> > Error: pkg/adaptor/hypervisor/registry/register.go:9:9: undefined: newServer
> > make: *** [Makefile:59: vet] Error 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am confused a little bit for where is the newServer function?
> 
> https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/cmd/cloud-api-adaptor/main.go#L194, this code may helps. Maybe fix could be https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/pkg/adaptor/hypervisor/registry/azure.go#L12

I suppose it may a code bug? not sure, but I just remove this function, as in fact we use build tag to control the different cloud provider... hence, I suppose for local development it can be removed safety and replace by return `nil`? 

Also nit fix as result of `go fmt`

Please help enable CI as I am 1st time contributor. :-)

**Comment by SamYuan1990** [comment]
Hi @wainersm and @stevenhorsman, could you please help review this pr?

File: Makefile
Comment (knrt10) [inline Makefile:43]: Can you please add a help comment here like other make targets? Context: https://github.com/confidential-containers/cloud-api-adaptor/blob/d9caf989940fd77b0d16401ced272d573f5d8760/Makefile#L20

File: pkg/adaptor/hypervisor/registry/register.go
Comment (knrt10) [inline pkg/adaptor/hypervisor/registry/register.go:9]: Can you please add this is a different commit? Also why is it returning nil?

**Review by knrt10 — CHANGES_REQUESTED** [review]
Some suggestions

File: pkg/adaptor/hypervisor/registry/register.go
Comment (SamYuan1990) [reply pkg/adaptor/hypervisor/registry/register.go:9]: the `newServer` function, in fact is useless. 
as you can see ref to @huoqifeng 's comment above. https://github.com/confidential-containers/cloud-api-adaptor/pull/322#issuecomment-1290744174

the `NewServer` function is invoked base on build tag, and if we go through build tag this `newServer` function is not been invoked. or even if this file is able to be deleted? which I am not sure, hence I keep this file.

the reason I make this line to return `nil`
- `hypervisor.Server` is an interface, as the `NewServer` function seems not been invoked according to build tag is not set, I suppose the default value of an interface is `nil`
- as this file without build tag, during go vet process it will try to find `newServer` function as original code. But we don't have any `newServer` function implemented without any build tag. which breaks vet check.

Comment (knrt10) [reply pkg/adaptor/hypervisor/registry/register.go:9]: Thank you for the explaining. Makes sense

File: Makefile
Comment (SamYuan1990) [reply Makefile]: updated, please review.

**Comment by SamYuan1990** [comment]
sorry for mistake as remove @wainersm from review list.

**Review by huoqifeng — DISMISSED** [review]
/lgtm and thanks @SamYuan1990 contribution!

**Comment by bpradipt** [comment]
/hold
Kindly test this change with different providers before merge.

File: pkg/adaptor/hypervisor/registry/register.go
Comment (bpradipt) [reply pkg/adaptor/hypervisor/registry/register.go:9]: @SamYuan1990 Looks like I'm missing something.
The `NewServer` method is the public method called from main. This invokes provider specific`newServer` based on build tag. If NewServer returns nil then how is it going to work?

Comment (huoqifeng) [reply pkg/adaptor/hypervisor/registry/register.go:9]: > @SamYuan1990 Looks like I'm missing something. The `NewServer` method is the public method called from main. This invokes provider specific`newServer` based on build tag. If NewServer returns nil then how is it going to work?

@bpradipt @SamYuan1990 I'm sorry I misunderstood it, right, it could not return nil. maybe, should we change registry to a factory and return concrete server in https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/cmd/cloud-api-adaptor/main.go#L194? -- I got the same error report when build the repo in my pipeline system.

Comment (bpradipt) [reply pkg/adaptor/hypervisor/registry/register.go:9]: @huoqifeng I would avoid changing it unless there is a good reason behind it.
Although to be honest, I don't understand this change :-( hence my question to @SamYuan1990 .
Btw, can you point me to the error that is reported ?

File: .github/workflows/lint.yaml
Comment (bpradipt) [inline .github/workflows/lint.yaml:40]: vsphere provider is missing

Comment (bpradipt) [inline .github/workflows/lint.yaml:40]: vsphere provider is missing

File: pkg/adaptor/hypervisor/registry/register.go
Comment (SamYuan1990) [reply pkg/adaptor/hypervisor/registry/register.go:9]: wait a min, @huoqifeng , @bpradipt , if this function return nil breaks pipeline for build, why the build pipeline with tag successes?

Comment (SamYuan1990) [reply pkg/adaptor/hypervisor/registry/register.go:9]: https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/pkg/adaptor/hypervisor/ibmcloud/service.go#L55 em... will update this pr.

Comment (SamYuan1990) [reply pkg/adaptor/hypervisor/registry/register.go]: @huoqifeng ， @bpradipt so far, I committed with a mock implementation which return nil when build tag is not given as an implementation of `newServer` to avoid error happen as go vet will not read build tag, as
```
# github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/registry
Error: pkg/adaptor/hypervisor/registry/register.go:9:9: undefined: newServer
make: *** [Makefile:59: vet] Error 2
```

File: .github/workflows/lint.yaml
Comment (SamYuan1990) [reply .github/workflows/lint.yaml:40]: added

File: pkg/adaptor/hypervisor/registry/register.go
Comment (huoqifeng) [reply pkg/adaptor/hypervisor/registry/register.go]: > @huoqifeng I would avoid changing it unless there is a good reason behind it. Although to be honest, I don't understand this change :-( hence my question to @SamYuan1990 . Btw, can you point me to the error that is reported ?

@bpradipt my company requires code scan before using it, the scan reported similar errors. Let me create an issue to address those issues reported separately.

Comment (SamYuan1990) [reply pkg/adaptor/hypervisor/registry/register.go]: thanks @huoqifeng , I just updated code as adding build tag in go vet process. as you mentioned at https://github.com/confidential-containers/cloud-api-adaptor/issues/327

**Review by huoqifeng — APPROVED** [review]
/lgtm

**Comment by bpradipt** [comment]
Thanks @SamYuan1990..This looks much safer now and we can run it and gradually fix the issues without breaking the working code.

@wainersm @stevenhorsman @knrt10 please take a look and provide your ack so that we can move this forward

**Review by knrt10 — APPROVED** [review]
LGTM

**Pull request merged**
